### PR TITLE
 Handle data collected beyond end of a frame

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
# Fixes 15825

LARMOR and SANS2D are starting to operate in two different time regimes for monitors and detectors. The data shows frame spillage, i.e. monitors further down the beam line measure data from the previous frame.

The problem is solved (for linearly spaced monitor workspaces) by duplicating the data set and allowing the user to specify the allowed wavelength range.

**To test:**
- Load the file LARMOR00009848_corrected.nxs which you can find here: /olympic/babylon5/Scratch/Anton/Testing/15825
  - Plot the monitors (1-10) and confirm that there is chopped up data for some of them (spectrum 4). This is best seen by setting the axes to log-log.
- Apply the `UnwrapMonitorsInTOF` algorithm with only the workspace set
  - Confirm that the output workspace has now twice the TOF range and the data is appended to itself
- Apply the `UnwrapMonitorsInTOF` algorithm with setting the range to 2 to 15
  - Confirm that for monitor 4 only the central (physical sensible) bit is left. This is best seen by setting the axes to log-log.

**Release Notes**

See [here](https://github.com/mantidproject/mantid/pull/17371/commits/a963c8ed0dab75d738dd84ef2ad12486a73b3bfa)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/inc/MantidAlgorithms/UnwrapMonitorsInTOF.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#include "MantidAPI/MatrixWorkspace_fwd.h"
+
+namespace Mantid {
+namespace Algorithms {
+
+/** UnwrapMonitorsInTOF : Handles workspaces which only contain monitors
+ *  and have data recorded which spills over from the previous frame.
+ *  This can occur when dealing with different time regimes for detectors
+ *  and monitors.
+
```
Comment (ianbush) [inline Framework/Algorithms/inc/MantidAlgorithms/UnwrapMonitorsInTOF.h]: This mentions workspaces that only contain monitors, but I think it checks and only moves the monitors.

I don't think the tests check that only monitors are moved, perhaps that should be verified too?

File: Framework/Algorithms/src/UnwrapMonitorsInTOF.cpp [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+};
+
+/**
+ * For a given distance from the source and lower and upper wavelength bound the
+ *function calculates the lower and upper bound of the
+ * TOF for this distance.
+ *
+ * To get the time: T = L/V and V = h/(m*lambda) --> T(lambda) = (L*m/h)*lambda
+ * In addition we need to divide the wavelength by 10^10 since the input is in
+ *Anstrom and we need to multiply by 10^6 since the
```
Comment (ianbush) [inline Framework/Algorithms/src/UnwrapMonitorsInTOF.cpp]: Typo

File: docs/source/algorithms/UnwrapMonitorsInTOF-v1.rst [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      dataY[index] = y_values [index]
+
+
+  # Apply the UnwrapMonitorsInTOF algorithm
+  output_workspace = UnwrapMonitorsInTOF(InputWorkspace=workspace, WavelengthMin=5, WavelengthMax=20)
+
+  # Inspect the unwrapped data
+  dataY_doubled = output_workspace.dataY(3)
+  print("The number of bins is {0} and is expected to be 20.".format(len(dataY_doubled)))
+  print("The 1st entry is {0}.".format(dataY_doubled[0]))
```
Comment (ianbush) [inline docs/source/algorithms/UnwrapMonitorsInTOF-v1.rst]: I thought it was clearer what was going on when you print the entire array dataY. I don't know if this looks good though or plays nicely with the doc tests.

**Comment by ianbush** [comment]
Tested the described functional test, and everything works as described. Very clear user documentation! A few of minor comments added but otherwise looks good.

**Comment by AntonPiccardoSelg** [comment]
@ianbush 
I have made some changes which addressed:
* Misleading algorithm description
* Test for non-monitor spectra
* Typo
* Full array output for doc test

**Comment by ianbush** [comment]
Looks good. :shipit: once Jenkins gives the :+1:.

**Pull request merged**